llvm/lib/Target/AIE/AIE2AliasAnalysis.cpp

```diff
@@ -0,0 +1,68 @@
+//===- AIE2AliasAnalysis.cpp ----------------------------------------------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+/// \file
+/// This is the AIE alias analysis pass. It looks through AIE's complex
+/// addressing mode intrinsics to find the underlying base object, then
+/// calls back into regular alias analysis to infer NoAlias relations.
+//===----------------------------------------------------------------------===//
+
+#include "AIE2AliasAnalysis.h"
+#include "AIE2.h"
+#include "AIE2AddrSpace.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/MemoryLocation.h"
+#include "llvm/Analysis/ValueTracking.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicsAIE2.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+
+static cl::opt<bool>
+    ASBasedAA("aie2-AS-AA", cl::init(false),
+              cl::desc("Alias Analysis based on Address Space"), cl::Hidden);
+
+#define DEBUG_TYPE "aie2-aa"
+
+char AIE2AAWrapperPass::ID = 0;
+
+INITIALIZE_PASS(AIE2AAWrapperPass, "aie2-aa",
+                "AIE complex addressing modes based Alias Analysis", false,
+                true)
+
+ImmutablePass *llvm::createAIE2AAWrapperPass() {
+  return new AIE2AAWrapperPass();
+}
+
+AIE2AAWrapperPass::AIE2AAWrapperPass() : ImmutablePass(ID) {
+  initializeAIE2AAWrapperPassPass(*PassRegistry::getPassRegistry());
+}
+
+void AIE2AAWrapperPass::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.setPreservesAll();
+}
+
+AliasResult AIE2AAResult::alias(const MemoryLocation &LocA,
+                                const MemoryLocation &LocB, AAQueryInfo &AAQI,
+                                const Instruction *CtxI) {
+  if (ASBasedAA) {
```

Comment:
Since we always annotation for load & store to TM from the clang frontend. 
We can safely add `NoAlias` when one of the loc is having AS of TM and `MayAlias` if both loc is having AS of TM.